src/atlas/array/ArrayViewVariant.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * (C) Crown Copyright 2024 Met Office
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#pragma once
+
+#include <variant>
+
+#include "atlas/array.h"
+
+namespace atlas {
+namespace array {
+
+namespace detail {
+
+using namespace array;
+
+// Container struct for a list of types.
+template <typename... Ts>
+struct Types {
+  using add_const = Types<std::add_const_t<Ts>...>;
+};
+
+// Container struct for a list of integers.
+template <int... Is>
+struct Ints {};
+
+template <typename Values, typename Ranks, typename... ArrayViews>
+struct VariantHelper;
+
+// Recursively construct ArrayView std::variant from types Ts and ranks Is.
+template <typename T, typename... Ts, int... Is, typename... ArrayViews>
+struct VariantHelper<Types<T, Ts...>, Ints<Is...>, ArrayViews...> {
+  using type = typename VariantHelper<Types<Ts...>, Ints<Is...>, ArrayViews...,
+                                      ArrayView<T, Is>...>::type;
+};
+
+// End recursion.
+template <int... Is, typename... ArrayViews>
+struct VariantHelper<Types<>, Ints<Is...>, ArrayViews...> {
+  using type = std::variant<ArrayViews...>;
+};
+
+template <typename Values, typename Ranks>
+using Variant = typename VariantHelper<Values, Ranks>::type;
+
+}  // namespace detail
+
+/// @brief Supported ArrayView value types.
+using ValueTypes = detail::Types<float, double, int, long, unsigned long>;
+
+/// @brief Supported ArrayView ranks.
+using Ranks = detail::Ints<1, 2, 3, 4, 5, 6, 7, 8, 9>;
```

Comment:
I would prefer if these were in the detail namespace as well, or is this something you require access to in user code?

Reply:
Yeah, probably an oversight on my part. Makes more sense to have them _reasonably_ hidden in the  `detail` namespace.

Reply:
Done!